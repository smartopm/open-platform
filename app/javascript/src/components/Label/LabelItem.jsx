import React from 'react'
import {
  ListItem,
  Typography,
  IconButton, Grid
} from '@material-ui/core'
import MoreVertIcon from '@material-ui/icons/MoreVert'
import PropTypes from 'prop-types'
import { makeStyles } from "@material-ui/core/styles";


// shortDesc
// number of users
// color
// description
export default function LabelItem({ label, userType, userCount }) {
  // eslint-disable-next-line no-use-before-define
  const classes = useStyles();

  function handleOpenMenu() {
    // handle modal stuff here
  }
  return (
    <ListItem key={label.id} className={classes.labelItem}>
      <Grid container spacing={6}>
        <Grid item xs={3}>
          <Typography variant="subtitle1" data-testid="label-title">
            {label.shortDesc}
          </Typography>
        </Grid>
        <Grid item xs={3}>
          <Typography variant="subtitle1" data-testid="label-users">
            {userCount}
          </Typography>
        </Grid>
        <Grid item xs={3}>
          <Typography variant="subtitle1" data-testid="label-description">
            This is a label description
          </Typography>
        </Grid>
        <Grid item xs={3}>
          {userType === 'admin' && (
          <IconButton
            className={classes.menuButton}
            aria-label={`more-${label.shortDesc}`}
            aria-controls="long-menu"
            aria-haspopup="true"
            onClick={handleOpenMenu}
            dataid={label.id}
          >
            <MoreVertIcon />
          </IconButton>
      )}
        </Grid>
      </Grid>
    </ListItem>
  )
}

LabelItem.propTypes = {
    label: PropTypes.shape({
        id: PropTypes.string,
        shortDesc: PropTypes.string
    }).isRequired,
    userType: PropTypes.string.isRequired,
    userCount: PropTypes.number.isRequired
}

const useStyles = makeStyles(() => ({
  labelItem: {
      borderBottomStyle: 'solid',
      borderBottomColor: '#F6F6F6',
      borderBottom: 10,
      backgroundColor: '#FFFFFF'
  },
  menuButton: {
    float: 'right'
  },
}));