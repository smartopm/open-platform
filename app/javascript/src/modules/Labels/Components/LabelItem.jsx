import React, { useState } from 'react';
import { IconButton, Grid } from '@mui/material';
import { useTranslation } from 'react-i18next';
import PropTypes from 'prop-types';
import makeStyles from '@mui/styles/makeStyles';
import { MoreHorizOutlined } from '@mui/icons-material';
import LabelActionMenu from './LabelActionMenu';
import Label from '../../../shared/label/Label';
import DataList from '../../../shared/list/DataList';
import Text from '../../../shared/Text';

export default function LabelItem({ label, userType, refetch }) {
  const [anchorEl, setAnchorEl] = useState(null);
  const { t } = useTranslation('common');
  const anchorElOpen = Boolean(anchorEl);
  const classes = useStyles();
  const labelsHeader = [
    {
      title: 'Labels',
      value: t('table_headers.labels'),
      col: 2
    },
    { title: 'No of Users', value: t('table_headers.labels_total_no_of_users'), col: 2 },
    {
      title: 'Description',
      value: t('table_headers.labels_description'),
      col: 2
    },
    { title: 'Menu', value: t('common:table_headers.menu'), col: 1 }
  ];

  const menuData = {
    handleClick: event => setAnchorEl(event.currentTarget),
    anchorEl,
    open: anchorElOpen,
    handleClose: () => setAnchorEl(null),
    refetch
  };
  return (
    <>
      <div className={classes.planList}>
        <DataList
          keys={labelsHeader}
          data={renderLabels(label, menuData, userType)}
          hasHeader={false}
        />
      </div>
    </>
  );
}

export function renderLabels(label, menuData, userType) {
  return [
    {
      Labels: (
        <Grid item xs={12} md={2} data-testid="short_desc">
          <Label color={label.color} title={label.shortDesc} />
        </Grid>
      ),
      'No of Users': (
        <Grid item xs={12} md={2} data-testid="user_count">
          <Text content={label.userCount} />
        </Grid>
      ),
      Description: (
        <Grid item xs={12} md={2} data-testid="description">
          <Text content={label.description} />
        </Grid>
      ),
      Menu: (
        <Grid item xs={12} md={1} data-testid="menu">
          {userType === 'admin' && (
            <>
              <IconButton
                aria-controls="sub-menu"
                aria-haspopup="true"
                data-testid="label-menu"
                dataid={label.id}
                onClick={event => menuData.handleClick(event)}
                size="large"
              >
                <MoreHorizOutlined />
              </IconButton>
              <LabelActionMenu
                data={label}
                anchorEl={menuData.anchorEl}
                handleClose={menuData.handleClose}
                open={menuData.open && menuData?.anchorEl?.getAttribute('dataid') === label.id}
                refetch={menuData.refetch}
              />
            </>
          )}
        </Grid>
      )
    }
  ];
}

LabelItem.propTypes = {
  label: PropTypes.shape({
    id: PropTypes.string,
    shortDesc: PropTypes.string,
    color: PropTypes.string,
    description: PropTypes.string,
    userCount: PropTypes.number
  }).isRequired,
  userType: PropTypes.string.isRequired,
  refetch: PropTypes.func.isRequired
};

const useStyles = makeStyles(() => ({
  labelItem: {
    borderBottomStyle: 'solid',
    borderBottomColor: '#F6F6F6',
    borderBottom: 10,
    backgroundColor: '#FFFFFF'
  },
  labelGrid: {
    marginTop: '8px'
  },
  menuButton: {
    float: 'right'
  }
}));
