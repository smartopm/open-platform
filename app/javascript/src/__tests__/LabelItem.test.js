import React from 'react'
import { render } from '@testing-library/react'
import { MockedProvider } from '@apollo/react-testing'
import { BrowserRouter } from 'react-router-dom/'
import '@testing-library/jest-dom/extend-expect'
import LabelItem from '../components/Label/LabelItem'

describe('Label Item Component', () => {
  it('it should include the label details ', () => {
      const props = {
        label: {
            id: "2b3f902b-eb44",
            shortDesc: "com_news_sms",
            userCount: 10
        },
        userType: 'admin',
      }
    const container = render(
      <MockedProvider>
        <BrowserRouter>
          <LabelItem {...props} />
        </BrowserRouter>
      </MockedProvider>)
    expect(container.queryByText('com_news_sms')).toBeInTheDocument()
    expect(container.queryByText('10')).toBeInTheDocument()
    expect(container.queryByTestId('label-title')).toHaveTextContent('com_news_sms')
  })
})
