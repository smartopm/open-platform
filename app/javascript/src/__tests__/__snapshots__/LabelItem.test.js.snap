// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Label Item Component it should include the label details  1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <li
        class="MuiListItem-root makeStyles-labelItem-1 MuiListItem-gutters"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-6"
        >
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1"
              data-testid="label-title"
            >
              com_news_sms
            </h6>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1"
              data-testid="label-users"
            >
              10
            </h6>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle1"
              data-testid="label-description"
            >
              This is a label description
            </h6>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
          >
            <button
              aria-controls="long-menu"
              aria-haspopup="true"
              aria-label="more-com_news_sms"
              class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2"
              dataid="2b3f902b-eb44"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiIconButton-label"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                  />
                </svg>
              </span>
              <span
                class="MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
      </li>
    </div>
  </body>,
  "container": <div>
    <li
      class="MuiListItem-root makeStyles-labelItem-1 MuiListItem-gutters"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-6"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1"
            data-testid="label-title"
          >
            com_news_sms
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1"
            data-testid="label-users"
          >
            10
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle1"
            data-testid="label-description"
          >
            This is a label description
          </h6>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-3"
        >
          <button
            aria-controls="long-menu"
            aria-haspopup="true"
            aria-label="more-com_news_sms"
            class="MuiButtonBase-root MuiIconButton-root makeStyles-menuButton-2"
            dataid="2b3f902b-eb44"
            tabindex="0"
            type="button"
          >
            <span
              class="MuiIconButton-label"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 8c1.1 0 2-.9 2-2s-.9-2-2-2-2 .9-2 2 .9 2 2 2zm0 2c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm0 6c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
                />
              </svg>
            </span>
            <span
              class="MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </li>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
